R doc 1818 conditional load (#1257)
ravendb/docs: Documentation/5.1/Raven.Documentation.Pages/client-api/session/loading-entities.dotnet.markdown

# Session: Loading Entities
---

{NOTE: }

There are severl methods with many overloads that allow users to download documents 
from the database and convert them to entities. This article will cover the following 
methods:  

- [Load](../../client-api/session/loading-entities#load)
- [Load with Includes](../../client-api/session/loading-entities#load-with-includes)
- [Load - multiple entities](../../client-api/session/loading-entities#load---multiple-entities)
- [LoadStartingWith](../../client-api/session/loading-entities#loadstartingwith)
- [ConditionalLoad](../../client-api/session/loading-entities#conditionalload)
- [Stream](../../client-api/session/loading-entities#stream)
- [IsLoaded](../../client-api/session/loading-entities#isloaded)

{NOTE/}

---

{PANEL:Load}

> 

Reviewer: add links to Including Counters and Including TimeSeries 
`client-api/session/counters/counters-and-other-features#including-counters`

The most basic way to load a single entity is to use one of the `Load` methods.

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_1_0@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_1_0_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/} 

| Parameter | Type | Description |
| ------------- | ------------- | ----- |
| **id** | `string` | Identifier of a document that will be loaded. |

| Return Type | Description |
| ------------- | ----- |
| `TResult` | Instance of `TResult` or `null` if a document with a given ID does not exist. |

### Example

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_1_1@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_1_1_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/} 

> {NOTE In 4.x RavenDB, only string identifiers are supported. If you are upgrading from 3.x, this is a major change, because in 3.x non-string identifiers are supported. /}

Reviewer: ```suggestion
{NOTE In 4.x RavenDB and up, only string identifiers are supported. If you are upgrading from 3.x, this is a major change, because in 3.x non-string identifiers are supported. /}
```

Author: Done

{PANEL/}

{PANEL:Load with Includes}

> 

Reviewer: add links to Including Counters and Including TimeSeries
`client-api/session/counters/counters-and-other-features#including-counters`

Author: Done

| **pageSize** | `int` | maximum number of documents that will be retrieved |
| **exclude** | `string` | pipe ('&#124;') separated values for which document IDs (after 'idPrefix') should **not** be matched ('?' any single character, '*' any characters) |
| **skipAfter** | `string` | skip document fetching until given ID is found and return documents after that ID (default: `null`) |

| Return Type | Description |
| ------------- | ----- |
| `TResult[]` | Array of entities matching given parameters. |
| `Stream` | Output entities matching given parameters as a stream. |

### Example I

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_4_1@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_4_1_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/}

### Example II

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_4_2@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_4_2_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/}

{PANEL/}

{PANEL: ConditionalLoad}

The `ConditionalLoad` method takes a document's [change vector](../../server/clustering/replication/change-vector), 

> and if this change vector matches the document's current change vector on the server 
> side, it is not loaded. If the change vectors do not match, the document is loaded.  

Reviewer: ```suggestion
if the entity is tracked in the session, the method returns the entity, if not it checks if the provided change vector matches the document's current change vector on the server 
side, if the change vector match it is not loaded and NotModified returned. If the change vectors do not match, the document is loaded.  
```

Author: Done

| Return Type | Description |
| ------------- | ----- |
| ValueTuple `(T Entity, string ChangeVector)` | If the given change vector and the server side change vector do not match, the method returns the requested entity and its current change vector.<br/>If the change vectors match, the method returns `default` as the entity, and the current change vector.<br/>If the specified document, the method returns only `default` without a change vector. |

### Example

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_7_1@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_7_1_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/} 

{PANEL/}

{PANEL:Stream}

Entities can be streamed from the server using one of the following `Stream` methods from the `Advanced` session operations.

{CODE-TABS}
{CODE-TAB:csharp:Sync loading_entities_5_0@ClientApi\Session\LoadingEntities.cs /}
{CODE-TAB:csharp:Async loading_entities_5_0_async@ClientApi\Session\LoadingEntities.cs /}
{CODE-TABS/}

| Parameter | Type | Description |
| ------------- | ------------- | ----- |
| **startsWith** | `string` | prefix for which documents should be streamed |

> | **matches** | `string | pipe ('&#124;') separated values for which document IDs should be matched ('?' any single character, '*' any characters) |

Reviewer: ```suggestion
| **matches** | `string` | pipe ('&#124;') separated values for which document IDs should be matched ('?' any single character, '*' any characters) |
```

Author: Thanks, fixed

| **start** | `int` | number of documents that should be skipped  |
| **pageSize** | `int` | maximum number of documents that will be retrieved |
| **skipAfter** | `string` | skip document fetching until a given ID is found and returns documents after that ID (default: `null`) |

> | `streamQueryStats` (out parameter) | Information about the streaming query (amount of results, which index was queried, etc.) |

Reviewer: move `Information about the streaming query (amount of results, which index was queried, etc.)` to the description column

Author: Fixed

| Return Type | Description |
| ------------- | ----- |

> | `IEnumerator<[StreamResult](../../glossary/stream-result)>` | Enumerator with entities. |

Reviewer: broken link

Author: Fixed
